### PR TITLE
Under-replicated Block Implementation & Update script

### PR DESCRIPTION
**Pull request by xiezhiyu01**

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -292,7 +301,7 @@ func (writeHandle *WriteHandle) Write(data []byte) error {
 		}
 	}()
 
-	// Step 3: Read the blocks from the DataNodes in parallel
+	// Step 3: Write the blocks to the DataNodes in parallel
```
Comment (ZhenbangYou) [inline client/client.go:310]: Thank you for fixing my comment bug

File: namenode/main.go [hunk 2]
```diff
@@ -250,11 +251,101 @@ func (server *NameNode) ReportBlock(blockReport common.BlockReport, success *boo
...
+						asyncRpcCall := dataNodeClient.Go("DataNode.CreateReplication", createReplicationRequest, &unused, nil)
+
+						select {
+						case <-asyncRpcCall.Done:
+							if asyncRpcCall.Error != nil {
+								slog.Error("Error during Create Replication request", "error", asyncRpcCall.Error)
+							} else {
+								// Update the storage info
+								inode.rwlock.Lock()
+								storageInfo.DataNodes = append(storageInfo.DataNodes, replicaEndpoints...)
```
Comment (ZhenbangYou) [inline namenode/main.go:355]: Let me mark this and consider it later. I got a warning "loop variable storageInfo captured by func literal" for `storageInfo`

File: namenode/main.go [hunk 3]
```diff
@@ -307,21 +400,32 @@ func (server *NameNode) pickDatanodes(num uint) []string {
 			break
 		}
 		for endpoint, _ := range iter.Value().(map[string]bool) {
```
Comment (ZhenbangYou) [inline namenode/main.go:433]: This can be simplified as `for endpoint := range iter.Value().(map[string]bool)`. I got this from IDE.

Comment (ZhenbangYou) [reply namenode/main.go:433]: Oh I guess this is my bad, let me fix it. Sorry for neglecting an IDE warning.

File: namenode/main.go [hunk 4]
```diff
@@ -250,11 +251,101 @@ func (server *NameNode) ReportBlock(blockReport common.BlockReport, success *boo
...
+	slog.Info("Replication monitor setup")
+
+	for {
+		time.Sleep(common.REPLICATION_MONITOR_INTERVAL) // Check periodically
+		slog.Info("Replication monitor running")
+
+		for fileName, inode := range server.inodes {
+			inode.rwlock.RLock()
+			for blockIndex, storageInfo := range inode.storageInfo {
+				if len(storageInfo.DataNodes) < common.BLOCK_REPLICATION && storageInfo.LatestVersion > 0 {
```
Comment (ZhenbangYou) [inline namenode/main.go]: I suggest replacing this with `storageInfo.LastestVersion >= common.MIN_VALID_VERSION_NUMBER`

File: namenode/main.go (hunk 2 above)
Comment (xiezhiyu01) [reply namenode/main.go:355]: Thanks! I'm not sure what should be the best practice here

Comment (ZhenbangYou) [reply namenode/main.go:355]: Lambda functions capture variables by reference, but what we want is capture by value. Thus, we have to pass this as a parameter of the lambda function.

**Pull request merged**